lib/spack/spack/cmd/stack.py

```diff
@@ -0,0 +1,120 @@
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+""" Stack spack installations - avoiding double-builds
+
+Stack this spack repository on top of an external read-only spack installation
+that already has several packages pre-built.
+
+The installed packages in the remote spack-installation are symlinked into the
+directory_layout of the current repository. Hardlinks make little to no sense
+since the original files need to remain in place because dependencies are found
+via RPATH (which are the old paths in the remote repository).
+"""
+
+from __future__ import print_function
+
+__author__ = "Oliver Breitwieser (UHEI)"
+
+import os
+import os.path as osp
+import spack
+import spack.config
+import spack.directory_layout
+import spack.store
+from spack.filesystem_view import filter_exclude
+from spack.util.path import canonicalize_path
+import llnl.util.tty as tty
+import llnl.util.filesystem as fs
+
+description = "integrate packages of remote spack repositories into this one"
+section = "admin"
+level = "long"  # TODO: re-check what 'level' is supposed to mean
+
+
+def setup_parser(sp):
+    setup_parser.parser = sp
+
+    sp.add_argument(
+        '-v', '--verbose', action='store_true', default=False,
+        help="If not verbose only warnings/errors will be printed.")
+    sp.add_argument(
+        '-e', '--exclude', action='append', default=[],
+        help="exclude packages with names matching the given regex pattern")
+
+    sp.add_argument("remotes", nargs="+",
+                    metavar='remote', action='store',
+                    help="one or more remote spack repositories which "
+                         "contain packages to be added as external packages")
+
+
+def add_remote_packages(remote, exclude=[], verbose=False):
+    """
+        Add all installed packages in `remote` to the packages dictionary.
+    """
+    config = spack.config.get_config("config")
+
+    # NOTE: This has to be kept in sync with spack/store.py!
+    layout = spack.directory_layout.YamlDirectoryLayout(
+        canonicalize_path(osp.join(remote, 'opt', 'spack')),
+        hash_len=config.get('install_hash_length'),
+        path_scheme=config.get('install_path_scheme'))
+
+    num_packages = 0
+
+    for spec in filter_exclude(layout.all_specs(), exclude):
+        src = layout.path_for_spec(spec)
+        tgt = spack.store.layout.path_for_spec(spec)
+        if osp.exists(tgt):
+            if osp.islink(tgt):
+                os.remove(tgt)
+            else:
```

Comment:
there should probably be a flag on how to handle this case. In general it should still be a valid stacked spack if it has some of the remote's packages locally, but this might indicate that something is wrong -> allow for raising an error here?

Reply:
Well, if the `hash` is the same it should be the same installed spec, just in a different location, so we can just change the link and if the file is installed locally we do not overwrite it and print a warning - what would you want the switch to do, exactly?

Reply:
basically just: always take existing link, always take new link